core/common/src/GenericInstanceFilter.ts

```diff
@@ -0,0 +1,190 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+/** @packageDocumentation
+ * @module Utils
+ */
+
+/**
+ * Generic instance filter that has all the necessary information to build filtering query.
+ * @beta
+ */
+export interface GenericInstanceFilter {
+  /** Single filter rule or multiple rules joined by logical operator. */
+  rules: GenericInstanceFilterRule | GenericInstanceFilterRuleGroup;
+  /**
+   * Information about related instances that has access to the properties used in filter.
+   * These can be used to create `JOIN` clause when building `ECSQL` query. Each related property
+   * used in rule will have `sourceAlias` that matches `GenericInstanceFilterRelatedInstanceDescription.alias`.
+   * If more than one property of the same related instance is used, they will share the same alias.
+   */
+  relatedInstances: GenericInstanceFilterRelatedInstanceDescription[];
+  /**
+   * List of class names whose properties are used in rules. Might be used to find common base class when building
+   * filter for instances of different classes.
+   */
+  propertyClassNames: string[];
+  /**
+   * List of class names which will be used for additionally only querying instances of specific classes.
+   */
+  filteredClassNames?: string[];
+}
+
+/**
+ * Type definition that describes operators supported by `GenericInstanceFilterRule`.
```

Comment:
Please use doc links throughout (`[[Thing]]` and `[Thing]($package)` instead of `` `Thing` ``).